### ISSUE TITLE
Stack overflow in `test/pass/x.ncl` with `cargo test`

### ISSUE THREAD

**Issue by litchipi** [tweag/nickel#502]
**Describe the bug**
After adding new features to the code (#470), some tests fails because their stack overflows.

**To Reproduce**
Checkout branch `timcer/typeenv` or the master after #470 is merged, and copy / paste some of the lines of `test/pass/contracts.ncl`.
Can also be done with master right now, but you'll need to copy / paste more of the lines.
Will fail on every test that is constructed as such (taken from `test/pass/basics.ncl`):
```
let Assert = fun l x => x || %blame% l in

(1+1 == 2 | #Assert) &&
(1-2+3-4 == -2 | #Assert) &&
(2-3-4 == -5 | #Assert) &&
(-1-2 == -3 | #Assert) &&
(2*2 + 2*3 - 2*4 == 2 | #Assert) &&
(...)
(let x = 3 in (switch { `foo => 1, _ => x} (3 + 2)) == 3) &&

true
```

Then perform a `cargo test`. Note that using `nickel -f` may not fail as it allocates more stack.

**Expected behavior**
No stack overflow

**Environment**
 - Ubuntu 21.10, with Linux 5.13.0-22
 - Can be reproduced with a lot of copy / paste on `master`
 - Can be reproduced with few copy / paste after PR #470

**Additional context**
For a normal evaluation, 8Mib of stack is allocated, whereas for `cargo test` it only allocates 4Mib.
The problem comes from the recursive functions in the code, that adds a lot to the stack.
The reason is that big `A && B && C && D` create a long long tree branch that gets handled in the code with
much more stack usage than the same number of nodes dispersed in a balanced tree.

**Fix proposed**
- For a quick fix in order to handle this rapidly, rewrite the `test/pass/*.ncl` tests
- For a full fix, rewrite parts of the codes that are recursive into a looping process.

**Comment by litchipi** [comment]
Note: is maybe important enough to have its place in the pre-release tracking issue (#462)

**Comment by steshaw** [comment]
@yannham The PR referred to here (https://github.com/tweag/nickel/pull/470) was not merged. However, it refers to a problem of stack growth on `master`. Here is a Nickel file that blows the stack on `master` currently:

https://gist.github.com/steshaw/b5f25681795699c283d4f2c4a09df662

**Comment by steshaw** [comment]
As the type checker is a naturally recursive process, it occurs to me that we may not want to consider this a bug. The `stackoverflow.ncl` code above is quite unusual and constructed specifically to cause the problem. In practice, this kind of code would rarely occur except through a process of code generation.

What may be a good (enough) solution is to increase the stack size. I'm presuming that Nickel will typically be a short-run, single-threaded process. I can see that we already have some code to [increase the stack size to 4 MB during the tests](https://github.com/tweag/nickel/blob/c2e378f756f1b630c0f2449944f4d3b621fb5c9e/tests/pass.rs#L22-L33).

Currently, the release build of nickel does not core dump on `stackoverflow.ncl` (on my x86_64 Linux machine). Since Rust [does not determine the stack size of the main thread](https://doc.rust-lang.org/std/thread/#stack-size), we could the same tactic as the tests to set the stack size of the main thread by starting a new thread with an explicit size in `main()`.

Since this giant 1708 line [stack-check.ncl](https://gist.github.com/steshaw/c271eed2212fa7f5cb9e0c77c6f90e82) file runs successfully with a release build, there doesn't seem to be a need for immediate concern. If the stack size becomes a concern, it is easily remedied by increasing the default size. For tests in debug mode, we can bump the existing [STACK_SIZE](https://github.com/tweag/nickel/blob/396d837e1621366c30694bcd608aba1fb7aa43ef/tests/pass.rs#L9).

If it becomes required, we could make the stack size of the main thread configurable at runtime. For now, I suggest that we close this issue.

**Comment by yannham** [comment]
For the record, the typechecker could very well be generating constraints without growing the stack too much. But otherwise I agree with your arguments: it is an adversarial and unnatural example, and using recursive functions for the typechecker makes the code simple and maintainable, so it's fine to close for now. Will open a new issue if we hit this problem in an actual codebase.
